[RF] LikelihoodJob improved precision

egpbos opened the pull request:
# This Pull request:

## Changes or fixes:

Triggered by a failure to run an ATLAS Higgs combination fit using the MultiProcess-parallelized LikelihoodJob, this PR changes the way offsetting is handled in the TestStatistics::LikelihoodWrapper family of classes. This in turn improves precision of evaluations, making many parallelized likelihood evaluations now bit-wise exactly equal to non-parallel evaluations. This allows us to better debug the Higgs fit. After this PR, the public Higgs workspace used in rootbench can be fit with LikelihoodJob enabled.

In detail, this PR changes the following:

- Increased precision:
    * Per-component offsets: instead of one offset for the total LikelihoodWrapper, we switched to a vector of offsets: one for each likelihood component. This makes a difference only for RooSumL fits, i.e. simultaneous PDF fits or fits with constraint or global observable terms. This brings the results of these fits closer to the old-style RooNLLVar fits, because those also use per-component offsets (per-RooNLLVar in a RooAddition to be exact).
    * In LikelihoodJob::evaluate, the result_ KahanSum is no longer initialized to zero, but is initialized to the first value in the results_ array, both sum and carry term. This sometimes makes a difference: adding a term with a small but non-zero carry term to an existing sum with a zero sum and zero carry term can make the small non-zero carry term disappear.
    * Due to these changes and the earlier KahanSum updates, we were able to tighten the tolerance of tests in testLikelihoodSerial, testLikelihoodJob and testLikelihoodGradientJob, with many tests now passing EXPECT_EQ.
    * testLikelihoodGradientJob adds offsetting to the parameterized test matrices of the LikelihoodGradientJobTest cases to test all the above (and below) changes.

- Offset synchronization:
    * LikelihoodWrapper and LikelihoodGradientWrapper now store a shared_ptr to the offsets instead of raw offsets. At construction time within a MinuitFcnGrad, they get passed the same single offset object so that it is always kept synced between the different likelihood calculators. If it isn't synced and the likelihood gets different offsets at different times, the minimizer understandably gets very confused. This was the case before this commit, which was, in fact, a bug.
    * For LikelihoodJob and LikelihoodGradientJob, the offsets are also synchronized to all workers via update_state. For this, we simply check before evaluation whether offsets have changed since the previous evaluation and if so they are sent over the MultiProcess::Messenger. Note that while the LikelihoodGradientJob doesn't itself use the offset (it doesn't calculate the likelihood), it must still synchronize offsets, because during the gradient calculation the LikelihoodSerial objects on the workers are used to calculate the likelihoods there, so for them the offsets must be up to date.
    * The LikelihoodJob contains a LikelihoodSerial member as well now. This allows the LikelihoodJob to trigger calculating the offsets on the master process before sending them to the workers.
    * LikelihoodWrapper has some added private helper functions for managing offsets.

- Other miscellaneous changes:
    * LikelihoodWrapper::setApplyWeightSquared was implemented properly for RooSumL likelihoods as well, passing along the call to component RooUnbinnedLs. Note, however, that it is currently not reachable for users because there is no interface to pass this along from the RooMinimizer, which contains the MinuitFcnGrad, which contains the LikelihoodWrapper. A comment in MinuitFcnGrad.h refers to this, reminding future devs to also flip offsets_reset_ when (un)setting squared weights.
    * LikelihoodWrapper now holds the likelihood_type. This cleans up some code duplication in LikelihoodSerial and LikelihoodJob, which previously used the type only in their ctors, and avoids dynamic_casts later, e.g. on when calculating offsets.
    * A RooSubsidiaryL can now also be computed with LikelihoodSerial and LikelihoodJob; this case was still missing in their evaluation functions.
    * The LikelihoodSerial, LikelihoodJob and LikelihoodGradientJob "ConstrainedAndOffset" test cases used the wrong argument for the constrained parameter. These are now corrected from alpha_bkg_obs_A to become alpha_bkg_A.
    * In LikelihoodJobTest, the added test case "UnbinnedGaussian1DSelectedParameterValues" shows where splitting over events can lead to bit-wise differences. This test will be useful in the future if further precision enhancements are made.

## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

guitargeek commented:
Hi @egpbos, thanks for the PR! It looks good, but can you please update such that is passes on all CI nodes? Thanks!

egpbos commented:
Yes, will do. I have been busy fixing a bug and with some other obligations. I will get back to this soon.

guitargeek commented on roofit/roofitcore/src/TestStatistics/LikelihoodGradientJob.cxx:
```diff
@@ -37,9 +38,26 @@ LikelihoodGradientJob::LikelihoodGradientJob(std::shared_ptr<RooAbsL> likelihood
+   : LikelihoodGradientWrapper(std::move(likelihood), std::move(calculation_is_clean), N_dim, minimizer, std::move(offsets), std::move(offsets_save)),
+     grad_(N_dim)
+{
+   // Note to future maintainers: take care when storing the minimizer_fcn pointer. The
```
As of https://github.com/root-project/root/pull/12348, this comment is not true anymore. The `RooAbsMinimizerFcn` subclasses are now never cloned. In fact they can't even be cloned because the copy constructors are deleted.

I guess that means you can remove this comment here and also in other places? And probably you can simplify your code because you don't need to work around the `MinuitFcnGrad` being cloned?

egpbos replied:
Ah, I overlooked this comment. I accidentally almost already addressed it by removing all cloning and copying support. The only thing left was to remove the comments, which I now did as well (amended the previous commit).

egpbos commented:
I have rebased this commit on master of last Friday. Let's see whether it now passes CI. @guitargeek is it ok with you to postpone the changes from your review comment to a separate issue? It seems like a cleanup task not directly related to the purpose of this PR.

guitargeek commented:
@phsft-bot build

guitargeek commented:
Hi @egpbos, thank you so much for picking this up again!

Your `@phsft-bot build` command will not do anything though :slightly_smiling_face: We have retired the Jenkins CI and the bot.

Our new CI is ran automatically, and you can check it's status at the bottom of the PR thread.

Once are platforms are finished running, the GitHub actions bot will also post a comment with a link to a summary, where you can see at one glance which tests fail on which platform and how.

egpbos commented:
I figured as much, old habits :)

egpbos commented:
I think this can finally be reviewed, CI was passing (just removed a stray printf in the last force push).

egpbos commented:
The latest commit changes only formatting and now all CI checks pass (see the CI checks of the previous commit).

egpbos commented:
The ROOT CI / mac14 failure is not due to this PR (see the successful build in the previous commit).

guitargeek commented on roofit/roofitcore/inc/RooFit/TestStatistics/LikelihoodGradientWrapper.h:
```diff
@@ -37,12 +38,22 @@ class LikelihoodGradientWrapper {
    LikelihoodGradientWrapper(std::shared_ptr<RooAbsL> likelihood,
                              std::shared_ptr<WrapperCalculationCleanFlags> calculation_is_clean, std::size_t N_dim,
                              RooMinimizer *minimizer);
+   LikelihoodGradientWrapper(std::shared_ptr<RooAbsL> likelihood,
+                             std::shared_ptr<WrapperCalculationCleanFlags> calculation_is_clean, std::size_t N_dim,
+                             RooMinimizer *minimizer,
+                             std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets,
+                             std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets_save);
```
```suggestion
   LikelihoodGradientWrapper(std::shared_ptr<RooAbsL> likelihood,
                             std::shared_ptr<WrapperCalculationCleanFlags> calculation_is_clean, std::size_t N_dim,
                             RooMinimizer *minimizer,
                             std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets=nullptr,
                             std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets_save=nullptr);
```
I would suggest default arguments here, so that you can delete the original overload. Having separate definitions poses a risk that they go out of sync.

Also, can this constructor be made `private` to make it clear that it's not meant to be used directly? All instantiation should happen via `create()`, right?

egpbos replied:
Done and done.

guitargeek commented on roofit/roofitcore/inc/RooFit/TestStatistics/LikelihoodGradientWrapper.h:
```diff
@@ -37,12 +38,22 @@ class LikelihoodGradientWrapper {
    static std::unique_ptr<LikelihoodGradientWrapper>
    create(LikelihoodGradientMode likelihoodGradientMode, std::shared_ptr<RooAbsL> likelihood,
           std::shared_ptr<WrapperCalculationCleanFlags> calculationIsClean, std::size_t nDim, RooMinimizer *minimizer);
+   static std::unique_ptr<LikelihoodGradientWrapper>
+   create(LikelihoodGradientMode likelihoodGradientMode, std::shared_ptr<RooAbsL> likelihood,
+          std::shared_ptr<WrapperCalculationCleanFlags> calculationIsClean, std::size_t nDim, RooMinimizer *minimizer,
+          std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets,
+          std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets_save);
```
```suggestion
   static std::unique_ptr<LikelihoodGradientWrapper>
   create(LikelihoodGradientMode likelihoodGradientMode, std::shared_ptr<RooAbsL> likelihood,
          std::shared_ptr<WrapperCalculationCleanFlags> calculationIsClean, std::size_t nDim, RooMinimizer *minimizer,
          std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets=nullptr,
          std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets_save=nullptr);
```
Same suggestion as for the constructor.

egpbos replied:
Done.

guitargeek commented on roofit/roofitcore/inc/RooFit/TestStatistics/LikelihoodWrapper.h:
```diff
@@ -60,11 +60,20 @@ class LikelihoodWrapper {
 public:
    LikelihoodWrapper(std::shared_ptr<RooAbsL> likelihood,
                      std::shared_ptr<WrapperCalculationCleanFlags> calculation_is_clean);
+   LikelihoodWrapper(std::shared_ptr<RooAbsL> likelihood,
+                     std::shared_ptr<WrapperCalculationCleanFlags> calculation_is_clean,
+                     std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets,
+                     std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets_save);
```
```suggestion
   LikelihoodWrapper(std::shared_ptr<RooAbsL> likelihood,
                     std::shared_ptr<WrapperCalculationCleanFlags> calculation_is_clean,
                     std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets=nullptr,
                     std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets_save=nullptr);
```

guitargeek commented on roofit/roofitcore/inc/RooFit/TestStatistics/LikelihoodWrapper.h:
```diff
@@ -60,11 +60,20 @@ class LikelihoodWrapper {
 
    static std::unique_ptr<LikelihoodWrapper> create(LikelihoodMode likelihoodMode, std::shared_ptr<RooAbsL> likelihood,
                                                     std::shared_ptr<WrapperCalculationCleanFlags> calculationIsClean);
+   static std::unique_ptr<LikelihoodWrapper>
+   create(LikelihoodMode likelihoodMode, std::shared_ptr<RooAbsL> likelihood,
+          std::shared_ptr<WrapperCalculationCleanFlags> calculationIsClean,
+          std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets,
+          std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets_save);
```
```suggestion
   static std::unique_ptr<LikelihoodWrapper>
   create(LikelihoodMode likelihoodMode, std::shared_ptr<RooAbsL> likelihood,
          std::shared_ptr<WrapperCalculationCleanFlags> calculationIsClean,
          std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets=nullptr,
          std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets_save=nullptr);
```

guitargeek commented on roofit/roofitcore/src/TestStatistics/LikelihoodJob.h:
```diff
@@ -69,14 +76,16 @@ class LikelihoodJob : public MultiProcess::Job, public LikelihoodWrapper {
    std::size_t getNComponentTasks();
+
+   std::vector<ROOT::Math::KahanSum<double>> offsets_previous_;
+   std::unique_ptr<LikelihoodSerial> likelihood_serial_;
```
```suggestion
   LikelihoodSerial likelihood_serial_;
```
Why you're using a pointer here? In the constructor, the type looks to be clear at compile time:
```c++
     likelihood_serial_(std::make_unique<LikelihoodSerial>(likelihood_, calculation_is_clean_, component_offsets_,
                                                           component_offsets_save_))
```
Then, you also don't have to define the copy constructor manually, which is very error-prone if data members are added.

egpbos replied:
Replaced the pointer with a member.

Regarding copying: I decided to forbid it completely. It's not currently used and I don't think it should be. If the MinuitFcnGrad ever is copied, it should probably just create new Likelihood(Gradient)Wrappers. In any case, for the time being they are superfluous and as you say error prone. I basically just don't want to introduce them right now for the reason you mention: keeping them up to date will be a hassle (they are not trivial due to the shared offset object... what does copying mean? Would a user want it with a new offset or with the old one? I don't even think I can know, only the manager would know, like MinuitFcnGrad). So let's just leave them out to make sure when ever they do become necessary someone will make sure they will be fit for purpose at that time.

guitargeek commented on roofit/roofitcore/src/TestStatistics/LikelihoodWrapper.cxx:
```diff
@@ -49,13 +51,40 @@ namespace TestStatistics {
+                                     std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> offsets_save)
+   : likelihood_(std::move(likelihood)),
+     calculation_is_clean_(std::move(calculation_is_clean)),
+     component_offsets_(std::move(offsets)),
+     component_offsets_save_(std::move(offsets_save))
```
```suggestion
     component_offsets_(offsets ? std::move(offsets) : std::make_shared<std::vector<ROOT::Math::KahanSum<double>>>()),
     component_offsets_save_(offsets_save ? std::move(offsets_save) : std::make_shared<std::vector<ROOT::Math::KahanSum<double>>>())
```
Note that my other suggestions with avoiding the code duplication would mean something like this. Or, you put the `=make_shared...` in the declaration instead of `=nullptr`.

egpbos replied:
See comment below: no more default parameters, so this won't be necessary. In fact, nullptr shared_ptrs can still be passed explicitly, but the solution is not to handle that with a default value internally, because they always need to be shared.

guitargeek commented on roofit/roofitcore/inc/RooFit/TestStatistics/LikelihoodGradientWrapper.h:
```diff
@@ -82,6 +93,8 @@ class LikelihoodGradientWrapper {
    std::shared_ptr<RooAbsL> likelihood_;
    RooMinimizer *minimizer_;
    std::shared_ptr<WrapperCalculationCleanFlags> calculation_is_clean_;
+   std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> component_offsets_;
+   std::shared_ptr<std::vector<ROOT::Math::KahanSum<double>>> component_offsets_save_;
```
```suggestion
public:
   using OffsetVec = std::vector<ROOT::Math::KahanSum<double>>;

protected:
   std::shared_ptr<OffsetVec> component_offsets_;
   std::shared_ptr<OffsetVec> component_offsets_save_;
```
That's a very long type... Can this maybe be aliased? Then it would also be more bearable to use `std::make_shared<OffsetVec>()` as a default parameter value.

egpbos replied:
Yes, I will make this more readable.

egpbos replied:
I made an encapsulating class `SharedOffset` that manages the two `shared_ptr`s and also has your suggestion of `OffsetVec` alias, which in turn is also used elsewhere to make the code more readable everywhere.

egpbos commented:
@guitargeek While addressing your comments, I started thinking that there's perhaps a better design.

1. I don't think we should have a default parameter at all, but only one ctor with the offset parameters. In real life current cases, these classes are always used from MinuitFcnGrad and they always share offsets with one or two other Likelihood* objects. The only cases where no shared offsets are necessary is in unit tests, but in those cases it's easy to just create a dummy offset vector.
2. All the offset management is distracting a bit from the core Likelihood logic. I think it may be nicer to make a simple wrapper class `LikelihoodOffset` that holds the shared_ptrs to the vectors, does the swapping, and a few other things. This also makes it easier to create, because the unwieldy types that you also rightly point out would then only be contained within `LikelihoodOffset`.

I will start working on point 1 at least, and in the meantime will consider point 2.

egpbos commented:
I have also started on point 2 now.

egpbos commented:
Trying a rebase on master to make the CI pass on clang-format (✅) and mac14 which also mysteriously fails (I think it's not due to this PR, but still).

egpbos commented:
Failing test on ubuntu20 (1697:roottest-python-numba-numba) is unrelated to this PR.

The mac14 run already fails on configuration (some LibXml2 path failure). I don't see how this could be caused by this PR, but it's strange that this failure doesn't seem to be present in other PRs...

egpbos commented:
Ah, I see this PR also has the LibXml2 failure in the mac14 workflow: https://github.com/root-project/root/actions/runs/8956800442/job/24599041512

So indeed, CI failures unrelated to this PR.

egpbos commented:
I overlooked the review comment of Feb last year above. I accidentally almost already addressed it by removing all cloning and copying support. The only thing left was to remove the comments, which I now did as well. I amended the previous commit with that. The only difference with before is the removed comments, so behaviorally no change.

guitargeek reviewed (approved):
Thank you very much for the PR and addressing the review comments so thoroughly! I'll take care of the backport to the 6.32 branch, so this makes into May's release.
